webrender/src/prim_store.rs

```diff
@@ -257,15 +252,119 @@ pub struct RadialGradientPrimitiveGpu {
+        }
+    }
+}
+
+impl Clone for GradientData {
+    fn clone(&self) -> GradientData {
+        GradientData {
+            colors: self.colors,
+        }
+    }
+}
+
+impl GradientData {
+    // Generate a color ramp between the start and end indexes from a start color to an end color.
+    fn fill_colors(&mut self, start_idx: usize, end_idx: usize, start_color: &ColorF, end_color: &ColorF) -> usize {
+        if start_idx >= end_idx {
+            return start_idx;
+        }
+
+        // Calculate the color difference for individual steps in the ramp.
+        let inv_steps = 2.0 / (end_idx - start_idx) as f32;
+        let step_r = (end_color.r - start_color.r) * inv_steps;
+        let step_g = (end_color.g - start_color.g) * inv_steps;
+        let step_b = (end_color.b - start_color.b) * inv_steps;
+        let step_a = (end_color.a - start_color.a) * inv_steps;
+
+        let mut cur_idx = start_idx;
+        let mut cur_color = *start_color;
+        let mut cur_packed_color = PackedColor::from_color(&cur_color).to_bgra();
+
+        // Walk the ramp writing start and end colors for each entry.
+        while cur_idx < end_idx {
+            self.colors[cur_idx] = cur_packed_color;
+
+            cur_color.r += step_r;
+            cur_color.g += step_g;
+            cur_color.b += step_b;
+            cur_color.a += step_a;
+            cur_packed_color = PackedColor::from_color(&cur_color).to_bgra();
+            self.colors[cur_idx + 1] = cur_packed_color;
+
+            cur_idx += 2;
+        }
+
+        cur_idx
+    }
+
+    // Compute an entry index based on a gradient stop offset.
+    #[inline]
+    fn get_index(offset: f32) -> usize {
+        2 * (offset * GRADIENT_DATA_RESOLUTION as f32).round() as usize
+    }
+
+    // Build the gradient data from the supplied stops, reversing them if necessary.
+    fn build(&mut self, src_stops: &[GradientStop], reverse_stops: bool) {
+        let mut cur_idx = 0usize;
+        let mut cur_color = if let Some(src) = src_stops.first() {
+            src.color
+        } else {
+            ColorF::new(0.0, 0.0, 0.0, 0.0)
+        };
+
+        if reverse_stops {
+            // If the gradient is reversed, then ensure the stops are processed in reverse order
+            // and that the offsets are inverted.
+            for src in src_stops.iter().rev() {
+                cur_idx = self.fill_colors(cur_idx, Self::get_index(1.0 - src.offset),
+                                           &cur_color, &src.color);
+                cur_color = src.color;
+            }
+        } else {
+            for src in src_stops {
+                cur_idx = self.fill_colors(cur_idx, Self::get_index(src.offset),
+                                           &cur_color, &src.color);
+                cur_color = src.color;
+            }
+        }
+
+        // Fill out any remaining entries in the gradient.
+        self.fill_colors(cur_idx, 2 * GRADIENT_DATA_RESOLUTION, &cur_color, &cur_color);
```

Comment:
Nope, the inputs are already reversed before they get to fill_colors, so that everything from 0..N has already been filled. It's only potentially from N..end that we need to fill.